test/gtest/ucs/test_conn_match.cc

```diff
@@ -0,0 +1,212 @@
+
+#include <common/test.h>
+extern "C" {
+#include <ucs/datastruct/conn_match.h>
+}
+
+#include <vector>
+
+class test_conn_match : public ucs::test {
+public:
+    typedef struct {
+        ucs_conn_match_elem_t conn_match;
+        bool                  is_exp;
+        void                  *dest_address;
+        ucs_conn_sn_t         conn_sn;
+    } conn_elem_t;
+
+private:
+    conn_elem_t *retrieve_common(ucs_conn_match_elem_t *conn_match,
+                                 const void *dest_address,
+                                 ucs_conn_sn_t conn_sn, bool is_exp) {
+        if (conn_match == NULL) {
+            return NULL;
+        }
+
+        conn_elem_t *conn_elem = ucs_container_of(conn_match, conn_elem_t,
+                                                  conn_match);
+        EXPECT_EQ(is_exp,  conn_elem->is_exp);
+        EXPECT_EQ(conn_sn, conn_elem->conn_sn);
+        EXPECT_TRUE(!memcmp(dest_address, conn_elem->dest_address,
+                            m_conn_match_ctx.address_length));
+        return conn_elem;
+    }
+
+    void insert_common(const void *dest_address, ucs_conn_sn_t conn_sn,
+                       bool is_exp, conn_elem_t &elem) {
+        elem.is_exp       = is_exp;
+        elem.conn_sn      = conn_sn;
+        memcpy(elem.dest_address, dest_address,
+               m_conn_match_ctx.address_length);
+    }
+
+protected:
+    static const void* get_address(const ucs_conn_match_elem_t *conn_match) {
+        return ucs_container_of(conn_match, conn_elem_t,
+                                conn_match)->dest_address;
+    }
+
+    static ucs_conn_sn_t get_conn_sn(const ucs_conn_match_elem_t *conn_match) {
+        return ucs_container_of(conn_match, conn_elem_t,
+                                conn_match)->conn_sn;
+    }
+
+    static const char *address_str(const void *address, char *str,
+                                   size_t max_size) {
+        return ucs_strncpy_safe(str, static_cast<const char*>(address),
+                                ucs_min(m_conn_match_ctx.address_length,
+                                        max_size));
+    }
+
+    void conn_match_init(size_t address_length) {
+        ucs_conn_match_ops_t conn_match_ops;
+
+        conn_match_ops.get_address = get_address;
+        conn_match_ops.get_conn_sn = get_conn_sn;
+        conn_match_ops.address_str = address_str;
+
+        ucs_conn_match_init(&m_conn_match_ctx, address_length,
+                            &conn_match_ops);
+    }
+
+    void conn_match_cleanup() {
+        ucs_conn_match_cleanup(&m_conn_match_ctx);
+    }
+
+    void insert_exp(const void *dest_address, ucs_conn_sn_t conn_sn,
+                    conn_elem_t &elem) {
+        ucs_conn_match_insert(&m_conn_match_ctx, dest_address, conn_sn,
+                              &elem.conn_match, UCS_CONN_MATCH_QUEUE_EXP);
+        insert_common(dest_address, conn_sn, true, elem);
+    }
+
+    void insert_unexp(const void *dest_address, ucs_conn_sn_t conn_sn,
```

Comment:
maybe use single func for `exp` and `unexp` with extra param?
the same is relevant for `retrieve` funcs